Entitlements

ahgittin opened the pull request:
add per-user entitlements and documentation for entitlements

aledsage commented on core/src/main/java/brooklyn/management/entitlement/PerUserEntitlementManager.java:
```diff
@@ -0,0 +1,97 @@
+    public <T> boolean isEntitled(EntitlementContext context, EntitlementClass<T> entitlementClass, T entitlementClassArgument) {
+        EntitlementManager entitlementInEffect = null;
+        if (context==null || context.user()==null) {
+            // no user means it is running as an internal process, always has root
```
I'd prefer if we could use some special context to indicate this, rather than null. Can we intercept earlier, e.g. in `Entitlements.getEntitlementContext()` then don't return null.

The reason is that treating null as super-user feels risky as the behaviour if there is a programming error which would otherwise cause a NPE!

ahgittin replied:
Agree, but that's a major change, beyond the scope of this PR

aledsage commented on core/src/main/java/brooklyn/management/entitlement/PerUserEntitlementManager.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    public void addUser(String user, EntitlementManager managerForThisUser) {
+        perUserManagers.put(user, managerForThisUser);
```
Is it sensible to accept nulls, or should we `put(checkNotNull(user, "user"), checkNotNull(managerForThisUser, "managerForThisUser"))`?

ahgittin replied:
done

aledsage commented on core/src/test/java/brooklyn/management/entitlement/AcmeEntitlementManagerTest.java:
```diff
@@ -21,132 +21,38 @@
-        mgmt = null;
+    @Test
+    public void testAnonUserHasAllPermissions() {
+        checkUserHasAllPermissions("anon");
```
Why does "anon" have all permissions? I don't see "anon" defined in `AcmeEntitlementManager`.

ahgittin replied:
`root` was the default. have changed `anon` to `otherAuthorizedUser` so it's less surprising

aledsage commented on core/src/test/java/brooklyn/management/entitlement/AcmeEntitlementManagerTestFixture.java:
```diff
@@ -0,0 +1,157 @@
+    ConfigBag configBag;
+    
+    public void setup(ConfigBag cfg) {
+        mgmt = new LocalManagementContextForTests(BrooklynProperties.Factory.newEmpty().addFrom(cfg));
```
Why not do this in `init` (i.e. in the `@BeforeMethod`)?

ahgittin replied:
not all tests need an `app`, and i think this way the config order is clearer

aledsage commented:
Looks good; only very minor comments - happy for this to be merged once you've looked through those.

ahgittin commented:
addressed comments.  however there are conflicts with the docs restructuring, so will hang file until #458 is in (which has a placeholder for the docs mentioned here)
